### ISSUE TITLE
Bug with Monogame SpriteFont Importer

### ISSUE THREAD

**Issue by BenMullenger** [MonoGame/MonoGame#3984]
Been having a problem with displaying Japanese characters in my app ever since using the monogame pipeline.

![help for charracters](https://cloud.githubusercontent.com/assets/11158981/8512806/23c806f8-234c-11e5-88e4-4beced7ac4e2.jpg)

 So I look and found that the spriteFonts are imported by a new importer. I put a stop point in my program before it loaded the font and swaped the Font xnb generated by the monogame importer for one made by the Xna one, this seem to solve the problem i was having. 

![help for charracters4](https://cloud.githubusercontent.com/assets/11158981/8512809/30ac48d4-234c-11e5-8e43-8d992f8dab09.jpg)

Thought I should post this in case anyone else was having problems.

**Comment by tomspilman** [comment]
Totally possible that somewhere in the font pipeline we're handling extended characters like this incorrectly.  It is something that shouldn't be too hard to unit test and figure out.

**Comment by Nezz** [comment]
I wasn't too active regarding my report of missing non-Latin characters, but I think it's the same issue at its core.
https://github.com/mono/MonoGame/issues/3286

**Comment by tomspilman** [comment]
Anyone have time to investigate this one?

**Comment by KonajuGames** [comment]
If no-one grabs this in the next 12 hours, I'll have a look when I get home tonight.

**Comment by KonajuGames** [comment]
It renders the characters fine if the TTF contains the glyphs for the given character ranges.  Here is a sample using Arial Unicode MS and some Hiragana glyphs.  The square is U3040, which was included in the character range but does not exist in the TTF.  The asterisk is the DefaultCharacter for glyphs not included in the character ranges.

![image](https://cloud.githubusercontent.com/assets/694943/10667473/e84080f4-791a-11e5-916e-c618d924ad4d.png)

Can you provide us with more information, such as the contents of the spritefont file and maybe even the TTF itself if it is not a standard Windows font.

**Comment by KonajuGames** [comment]
I can only guess that perhaps XNA might be falling back to a standard Windows font such as Arial Unicode MS if the specified TTF does not contain the glyphs.

**Comment by tomspilman** [comment]
@Jamezila - I know you are having a similar problem.  Can you share the details of your `.spritefont` to see if we can resolve this?

**Comment by Jamezila** [comment]
For what it's worth I actually wrote some code to render the Japanese glyphs out into a PNG (using SpriteBatch.DrawString in XNA) so I could load that as normal content and just use my own text renderer for it.

But here's the .SpriteFont I was using; again, totally works in XNA, but just renders squares for Japanese glyphs in Mono.

    <?xml version="1.0" encoding="utf-8"?>
    
    <XnaContent xmlns:Graphics="Microsoft.Xna.Framework.Content.Pipeline.Graphics">
      <Asset Type="Graphics:FontDescription">

    
    <FontName>Arial</FontName>

    
    <Size>35</Size>

    
    <Spacing>0</Spacing>

    
    <UseKerning>true</UseKerning>

    
    <Style>Regular</Style>

    
    

    
    <CharacterRegions>
      <CharacterRegion>
        <Start>&#32;</Start>
        <End>&#126;</End>
      </CharacterRegion>
      <CharacterRegion><Start>&#x3001;</Start><End>&#x3002;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3005;</Start><End>&#x3005;</End></CharacterRegion>
      <CharacterRegion><Start>&#x300C;</Start><End>&#x300D;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3041;</Start><End>&#x3042;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3044;</Start><End>&#x3044;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3046;</Start><End>&#x3046;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3048;</Start><End>&#x3048;</End></CharacterRegion>
      <CharacterRegion><Start>&#x304A;</Start><End>&#x3061;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3063;</Start><End>&#x3073;</End></CharacterRegion>
      <CharacterRegion><Start>&#x3075;</Start><End>&#x3076;</End></CharacterRegion>
      ....
      <CharacterRegion><Start>&#xFF08;</Start><End>&#xFF09;</End></CharacterRegion>
      <CharacterRegion><Start>&#xFF1A;</Start><End>&#xFF1A;</End></CharacterRegion>
      <CharacterRegion><Start>&#xFF1F;</Start><End>&#xFF1F;</End></CharacterRegion>
    </CharacterRegions>
  </Asset>
    </XnaContent>

**Comment by KonajuGames** [comment]
I did some more research into this, and Windows and TrueType do indeed support fallback fonts for rendering Unicode glyphs that are not included in the TTF.  In your case, Arial does not include glyphs for any CJK characters.  In this case, the TTF can specify a fallback font for certain Unicode character ranges so that the TTF does not have to include those glyphs itself.  The fallback for `Arial` is `Arial Unicode MS`, which is a superset of Arial because it contains not only the same glyphs as Arial but also most of the Unicode pages including all of CJK.  The MonoGame content build pipeline uses FreeType to render glyphs, but FreeType does not support the fallback fonts.  This is something that we could look into supporting, but I'm not sure how easy it is.

To fix your current situation now, you can edit your spritefont file to use `Arial Unicode MS` instead of `Arial`, and it will work as expected in both XNA and MonoGame.

**Comment by Jamezila** [comment]
Ok, that makes sense, thanks!

**Comment by danzel** [comment]
We could add a ```<FallbackFont>``` section to the SpriteFont definition maybe?

**Comment by KonajuGames** [comment]
Ideally we would use the information in the TTF itself, but I don't know how accessible this information is.  Adding a `<FallbackFont>` node is a viable alternative though.

**Comment by tomspilman** [comment]
> Ideally we would use the information in the TTF itself

Yeah... we should investigate the possibility of figuring that out first.

Still `<FallbackFont>` might be needed for non-Windows platforms... so we might need both paths possibly.

**Comment by CartBlanche** [comment]
This is partially fixed with the #8554, but @tomspilman is looking at implementing a fall-back font system, hence why I've punted this to 3.8.4 milestone.

**Comment by tomspilman** [comment]
The bug which was bad support of multi-byte characters in the font system is fixed.

However we need to implement a font fallback option.  See https://github.com/MonoGame/MonoGame/issues/9195.
